src/drivers/uavcan_v1/Actuators/EscClient.hpp

```diff
@@ -0,0 +1,177 @@
+/****************************************************************************
+ *
+ *   Copyright (C) 2014 PX4 Development Team. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+/**
+ * @file EscClient.hpp
+ *
+ * Client-side implementation of DS-15 specification ESC service
+ *
+ * Publishes the following UAVCAN v1 messages:
+ *   reg.drone.service.actuator.common.sp.Value8.0.1
```

Comment:
```suggestion
 *   reg.drone.service.actuator.common.sp.Vector{2,3,4,6,8,31}.0.1
```

If you are using Classic CAN, Vector31 would be expensive to publish because it takes 3 frames and requires computing/validating CRC. There are two ways to address this:

- Type-safe way: select the right type based on the number of outputs.
- Fast and simple way: manually serialize an array of `float16` into a payload buffer and hand it over to libcanard. Serialization can be managed using `nunavutSetF16()` from `nunavut/support/serialization.h`.

Reply:
> * Fast and simple way: manually serialize an array of `float16` into a payload buffer and hand it over to libcanard. Serialization can be managed using `nunavutSetF16()` from `nunavut/support/serialization.h`

Could you elaborate on how this would be done?  Let's say for example you want to send 5 values - how does the CanardRxSubscription work on the far side when you send/receive an array of 5 floats?

Reply:
> how does the CanardRxSubscription work on the far side when you send/receive an array of 5 floats?


You always subscribe using Vector31. The first five elements will be read from the transfer and the rest will be zero-filled.

https://uavcan.org/specification/UAVCAN_Specification_v1.0-beta.pdf#page=40 *3.7.1.4 Implicit zero extension of missing data*

Reply:
Ok, I _thought_ that feature existed, but didn't know how to make use of it.  I'll take a closer look at Nunavut (or, maybe I'll leave that particular upgrade to someone else).